File: Params.java
public class Params {
    public static void doit1(int a) {
        a = 19;
    }
    
    public static void doit2(int[] arr) {
        arr[0] = 47;
    }

    public static void doit3(int[] arr) {
        arr = new int[5];
    }

    public static void main(String[] args) {
        int x = 12;
        doit1(x);
        System.out.println(x);

        int[] array = {1,2,3};
        doit2(array);
        for (int i=0; i < array.length; i++)
            System.out.println(array[i]);
    }
}

File: params.cpp
#include <stdio.h>

// call by value
void doit1(int x) {
    x = 12;
}


// call by value
void doit2(int *x) {
    *x = 12;
}


// call by reference (C++ only, not C)
//void doit3(call-by-reference int x)
//void doit3(ref int x)
//void doit3(int ref x)
void doit3(int &x) {
    printf("address of x%p\n",&x);
    x = 19;
}



int main() {
    int y = 9;
    doit1(y);
    // See! call by value.
    printf("%i\n",y);

    // Still call by value; difference is a
    // memory address is the value being copied
    // in
    doit2(&y);
    printf("%i\n",y);

    int z = 5;
    doit3(z);
    printf("%i\n",z);
    printf("address of z%p\n",&z);
    

    
    
}
    
